circuit-foundation/src/commonMain/kotlin/com/slack/circuit/foundation/SharedElementTransition.kt

```diff
@@ -0,0 +1,165 @@
+@Composable
+public fun SharedElementTransitionLayout(
+  modifier: Modifier = Modifier,
+  content: @Composable SharedElementTransitionScope.() -> Unit,
+) {
+  SharedTransitionLayout(modifier = modifier) {
+    val sharedTransitionScope = this
+    val animatedVisibilityScope = findAnimatedVisibilityScope()
+    CompositionLocalProvider(
+      LocalSharedTransitionScope provides sharedTransitionScope,
+      LocalTransitionAnimatedVisibilityScope provides animatedVisibilityScope,
+    ) {
+      SharedElementTransitionScope(sharedTransitionScope, animatedVisibilityScope, content)
+    }
+  }
+}
+
+/**
+ * [SharedElementTransitionScope] creates a [SharedElementTransitionScope] for the child layouts in
+ * [content]. Any child layout of [SharedElementTransitionScope] can then use the receiver scope
+ * [SharedElementTransitionScope] to create standard shared element or shared bounds transitions.
+ *
+ * @param content The child composable to be laid out.
+ */
+@ExperimentalSharedTransitionApi
+@OptIn(InternalCircuitApi::class)
+@Composable
+public fun SharedElementTransitionScope(
+  sharedTransitionScope: SharedTransitionScope = LocalSharedTransitionScope.current,
+  animatedVisibilityScope: AnimatedVisibilityScope = findAnimatedVisibilityScope(),
+  content: @Composable SharedElementTransitionScope.() -> Unit,
+) {
+  val scope =
+    remember(sharedTransitionScope) {
+      SharedElementTransitionScopeImpl(animatedVisibilityScope, sharedTransitionScope)
+    }
+  with(scope) {
+    updateAnimatedScope(animatedVisibilityScope)
+    content()
+  }
+}
+
+/**
+ * [SharedElementTransitionScope] provides a [SharedTransitionScope] for the standard shared
+ * elements/shared bounds animations. This also provides a [AnimatedVisibilityScope], which can be
+ * set using [LocalTransitionAnimatedVisibilityScope]. Typically this should be set by a
+ * [NavDecoration] that has children using [SharedElementTransitionScope] for the animations to
+ * match up with the navigation animations.
+ */
+@OptIn(ExperimentalSharedTransitionApi::class)
+public interface SharedElementTransitionScope : SharedTransitionScope {
+  public val animatedVisibilityScope: AnimatedVisibilityScope
+
+  public fun isTransitionState(predicate: (EnterExitState) -> Boolean): Boolean {
+    return with(animatedVisibilityScope) {
+      predicate(transition.targetState) || predicate(transition.currentState)
+    }
+  }
+}
+
+/**
+ * Finds the [AnimatedVisibilityScope] from the [LocalTransitionAnimatedVisibilityScope] or creates
+ * a [SimpleAnimatedVisibilityScope].
+ */
+@Composable
+private fun findAnimatedVisibilityScope(): AnimatedVisibilityScope {
+  return LocalTransitionAnimatedVisibilityScope.current ?: staticAnimatedVisibilityScope()
+}
+
+/**
+ * Creates an [AnimatedVisibilityScope] that is always visible as a fallback for any
+ * [SharedElementTransitionScope].
+ */
+@Composable
+private fun staticAnimatedVisibilityScope(): AnimatedVisibilityScope {
```

Comment:
This feels like it could potentially introduce some UI inconsistencies, is it possible that a developer somehow relies on the scope to perform a different UI animations unrelated to shared elements? Maybe it'd be less error-prone to rather have a nullable scope instead. 

Or have the Nav framework itself always expose the AnimatedVisibilityScope and throw an exception if its not there as something would've gone quite wrong to have that happen?

Reply:
I second what Rebecca said. This custom scope will always have the target visibility be `EnterExitState.Visible`. That would potentially prevent the shared elements from identifying the target bounds. More specifically when both sharedElement/sharedBounds call sites of the same key report their target visibility to be visible, we would not know which bounds to use as the target bounds since they are both "entering".

Reply:
Yup, it's pretty broken always being in the single state right now. Going to lean on the nav for providing this.